### PR TITLE
Collection View

### PR DESCRIPTION
**Pull request by Oschly**
Let me know what You think about this implementation of Collection View.

### CODE REVIEW

File: Sources/SwiftUIKit/Views/TableView.swift [hunk 1]
```diff
@@ -8,7 +8,7 @@
 import UIKit
 
 @available(iOS 11.0, *)
-public protocol CellDisplayable {
+public protocol CellDisplayable: UIView {
```
Comment (0xLeif) [inline Sources/SwiftUIKit/Views/TableView.swift:11]: The only thing that should use `CellDisplayable` is the data.

example
```swift
struct CellData {
    let value = String
}

extension CellData: CellDisplayable {
    var cellID: String {
        SomeCell.ID
    }
}
```

File: Sources/SwiftUIKit/Views/CollectionView.swift [hunk 2]
```diff
@@ -8,7 +8,7 @@
 import UIKit
 
 @available(iOS 11.0, *)
-public typealias CollectionViewCell = DataIdentifiable & CellConfigurable & CellUpdatable & UICollectionViewCell
+public typealias CollectionViewCell = CellDisplayable & DataIdentifiable & CellConfigurable & CellUpdatable & UICollectionViewCell
```
Comment (0xLeif) [inline Sources/SwiftUIKit/Views/CollectionView.swift:11]: The cells don't need `CellDisplayable`

File: Sources/SwiftUIKit/Views/TableView.swift [hunk 3]
```diff
@@ -28,7 +28,7 @@ public protocol CellConfigurable {
 }
 
 @available(iOS 11.0, *)
-public typealias TableViewCell = DataIdentifiable & CellConfigurable & CellUpdatable & UITableViewCell
+public typealias TableViewCell = CellDisplayable & DataIdentifiable & CellConfigurable & CellUpdatable & UITableViewCell
```
Comment (0xLeif) [inline Sources/SwiftUIKit/Views/TableView.swift:31]: Ditto

File: Sources/SwiftUIKit/Views/CollectionView.swift [hunk 4]
```diff
@@ -525,3 +539,126 @@ extension CollectionView: UICollectionViewDelegate {
...
+    public func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
+        if let spacing = returnValidValue(for: minimumLineSpacingForSections, section: section) {
+            return spacing
+        }
+        
+        return 10
+    }
+    
+    public func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
+
```
Comment (0xLeif) [inline Sources/SwiftUIKit/Views/CollectionView.swift:644]: 🤔 

```swift
if let spacing = returnValidValue(for: minimumInteritemSpacingForSections, section: section) {
    return spacing
}
```

File: Sources/SwiftUIKit/Views/CollectionView.swift [hunk 5]
```diff
@@ -52,9 +52,23 @@ public class CollectionView: UICollectionView {
...
     
+    fileprivate var sectionsInsets: [UIEdgeInsets]? = nil
+    
+    fileprivate var minimumLineSpacingForSections: [CGFloat]? = nil
+    
+    fileprivate var minimumInteritemSpacingForSections: [CGFloat]? = nil
+    
+    fileprivate var headerSizesForSections: [CGSize]? = nil
+    
+    fileprivate var footerSizeForSections: [CGSize]? = nil
```
Comment (0xLeif) [inline Sources/SwiftUIKit/Views/CollectionView.swift:63]: Can we remove the extra lines, unless there are marks or comments

File: Sources/SwiftUIKit/Views/CollectionView.swift [hunk 6]
```diff
@@ -525,3 +539,126 @@ extension CollectionView: UICollectionViewDelegate {
...
+        }
+        
+        return self
+    }
+    
+    @discardableResult
+    func headerSize(shouldUpdate: Bool = true, forSections sizes: [CGSize]) -> Self {
+        headerSizesForSections = sizes
+        
+
```
Comment (0xLeif) [inline Sources/SwiftUIKit/Views/CollectionView.swift:583]: Extra newline

File: Sources/SwiftUIKit/Views/CollectionView.swift (hunk 4 above)
Comment (Oschly) [reply Sources/SwiftUIKit/Views/CollectionView.swift:644]: That's a method for handling given amount of spacings/anything in this extension - if array's count isn't equal to number of sections in collection view, it returns first value of that array. If it is equal, it returns size for given section index. If method unwrapped a nil array, it returns nil as well, so on the end I am adding default value in delegate method.

I should document that and maybe rename it to use more meaningful name.

If we don't want to handle that in this way, it won't be a problem to change it to something else.

File: Sources/SwiftUIKit/Views/CollectionView.swift [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+    
+    fileprivate var titles: [String]? = nil
+    fileprivate var sectionsInsets: [UIEdgeInsets]? = nil
+    fileprivate var footerSizeForSections: [CGSize]? = nil
+    fileprivate var headerSizesForSections: [CGSize]? = nil
+    fileprivate var minimumLineSpacingForSections: [CGFloat]? = nil
+    fileprivate var didSelectItemAtHandler: ((IndexPath) -> ())? = nil
+    fileprivate var canFocusItemAtHandler: ((IndexPath) -> Bool)? = nil
+    fileprivate var didDeselectItemAtHandler: ((IndexPath) -> ())? = nil
+    fileprivate var minimumInteritemSpacingForSections: [CGFloat]? = nil
```
Comment (0xLeif) [inline Sources/SwiftUIKit/Views/CollectionView.swift:25]: Can we group similar variables 

```swift
fileprivate var titles: [String]? = nil
fileprivate var sectionsInsets: [UIEdgeInsets]? = nil
fileprivate var footerSizeForSections: [CGSize]? = nil
fileprivate var headerSizesForSections: [CGSize]? = nil
fileprivate var minimumLineSpacingForSections: [CGFloat]? = nil
fileprivate var minimumInteritemSpacingForSections: [CGFloat]? = nil
```

File: Sources/SwiftUIKit/Views/CollectionView.swift (hunk 4 above)
Comment (0xLeif) [reply Sources/SwiftUIKit/Views/CollectionView.swift:633]: @Oschly I am fine with this. I was just pointing out this function doesn't have that bit of code I expect... It just returns 5 always. 

`minimumInteritemSpacingForSections` should be used in this function.

File: Sources/SwiftUIKit/Views/CollectionView.swift [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+    /// Returns value based on size of given array's count and `numberOfSections` integer.
+    /// - Parameters:
+    ///   - array: Array of objects containing value that can be returned
+    ///   - section: section which will be used to get value from array
+    /// - Returns: First value of array if array's count isn't equal `numberOfSections`, value from given index if array's count is equal to `numberOfSections` and nil if array is nil.
+    func returnValidValue<T>(for array: [T]?, section: Int) -> T? {
+            if let array = array,
+            array.count != 0 {
+            
+            if array.count != numberOfSections {
```
Comment (0xLeif) [inline Sources/SwiftUIKit/Views/CollectionView.swift:599]: This function doesn't check if `section > numberOfSections`

**Pull request merged**